File: internal/parser/stmt.go
package parser

import "github.com/obj-p/bazaar/internal/token"

type VarDeclStmt struct {
	Annotations []*AnnotationExpr `parser:"@@*"`
	Dest        *VariableExpr     `parser:"'var' @@"`
	Type        *TypeDecl         `parser:"@@?"`
	Source      *Expr             `parser:"'=' @@"`
}

type AssignStmt struct {
	Name  *string   `parser:"@Ident"`
	Op    *token.Op `parser:"@AssignOperator"`
	Value *Expr     `parser:"@@"`
}

type ForStmt struct {
	Dest   *VariableExpr `parser:"'for' ((@@"`
	Source *Expr         `parser:"'in' @@)"`
	Expr   *Expr         `parser:"| @@)"`
	Block  []*Stmt       `parser:"'{' @@* '}'"`
}

type ConditionalFragment struct {
	Var         *VarDeclStmt `parser:"@@"`
	ImplicitVar *string      `parser:"| 'var' @Ident"`
	Expr        *Expr        `parser:"| @@"`
}

type ConditionalBlock struct {
	Fragments []*ConditionalFragment `parser:"(@@ (',' @@)* ','?)"`
	Block     []*Stmt                `parser:"'{' @@* '}'"`
}

type IfStmt struct {
	If      *ConditionalBlock   `parser:"'if' @@"`
	ElseIfs []*ConditionalBlock `parser:"('else' 'if' @@)*"`
	Else    []*Stmt             `parser:"('else' '{' @@* '}')?"`
}

type SwitchCase struct {
	Expr  *Expr   `parser:"'case' @@ ':'"`
	Block []*Stmt `parser:"@@*"`
}

type SwitchBody struct {
	Cases   []*SwitchCase `parser:"@@*"`
	Default []*Stmt       `parser:"('default' ':' @@*)"`
}

type SwitchStmt struct {
	Expr *Expr       `parser:"'switch' @@"`
	Body *SwitchBody `parser:"'{' @@ '}'"`
}

type ReturnStmt struct {
	Value *Expr `parser:"'return' @@?"`
}

type CallStmt struct {
	Annotations    []*AnnotationExpr `parser:"@@*"`
	Name           *string           `parser:"@Ident"`
	Arguments      []*ArgumentExpr   `parser:"(('(' (@@ (',' @@)* ','?)? ')'"`
	TrailingLambda *LambdaExpr       `parser:"@@?)"`
	LambdaOnly     *LambdaExpr       `parser:"| @@)"`
}

type Stmt struct {
	Call   *CallStmt    `parser:"@@"`
	Var    *VarDeclStmt `parser:"| @@"`
	Assign *AssignStmt  `parser:"| @@"`
	For    *ForStmt     `parser:"| @@"`
	If     *IfStmt      `parser:"| @@"`
	Switch *SwitchStmt  `parser:"| @@"`
	Return *ReturnStmt  `parser:"| @@"`
	Expr   *Expr        `parser:"| @@"`
}

File: bazaar-parser/src/commonMain/kotlin/com/bazaar/parser/ast/Stmt.kt
package com.bazaar.parser.ast

sealed interface Stmt

// --- Simple statements ---

data class VarDeclStmt(
    val names: List<String>,
    val type: TypeDecl? = null,
    val value: Expr,
    val annotations: List<Annotation> = emptyList(),
) : Stmt

data class AssignStmt(val target: String, val op: AssignOp, val value: Expr) : Stmt

data class CallStmt(
    val expr: CallExpr,
    val annotations: List<Annotation> = emptyList(),
) : Stmt

data class ReturnStmt(val value: Expr? = null) : Stmt

data class ExprStmt(val expr: Expr) : Stmt

// --- If ---

data class IfStmt(
    val fragments: List<IfFragment>,
    val body: List<Stmt>,
    val elseIfs: List<ElseIf> = emptyList(),
    val elseBody: List<Stmt>? = null,
) : Stmt

data class ElseIf(val fragments: List<IfFragment>, val body: List<Stmt>)

sealed interface IfFragment

data class IfVarFragment(
    val names: List<String>,
    val type: TypeDecl? = null,
    val value: Expr? = null,
) : IfFragment

data class IfExprFragment(val expr: Expr) : IfFragment

// --- For ---

sealed interface ForStmt : Stmt

data class ForInStmt(
    val names: List<String>,
    val iterable: Expr,
    val body: List<Stmt>,
) : ForStmt

data class ForCondStmt(val condition: Expr, val body: List<Stmt>) : ForStmt

// --- Switch ---

data class SwitchStmt(
    val expr: Expr,
    val cases: List<SwitchCase>,
    val default: List<Stmt>? = null,
) : Stmt

data class SwitchCase(val expr: Expr, val body: List<Stmt>)
